rust/otap-dataflow/benchmarks/benches/retry_processor/main.rs

```diff
@@ -0,0 +1,331 @@
+//!
+//! This benchmark suite measures the performance characteristics of the retry processor
+//! under different load patterns and configurations. Key metrics include:
+//!
+//! 1. **Message Throughput**: Success path vs retry overhead
+//! 2. **HashMap Scalability**: Performance with large numbers of pending messages  
+//! 3. **Memory Cleanup**: Cost of expired message cleanup operations
+
+#![allow(missing_docs)]
+
+use criterion::{BenchmarkId, Criterion, Throughput, criterion_group, criterion_main};
+use mimalloc::MiMalloc;
+use otap_df_channel::mpsc;
+use otap_df_engine::retry_processor::{RetryConfig, RetryProcessor};
+use otap_df_engine::{
+    local::processor::{EffectHandler, Processor},
+    message::{ControlMsg, Message, Sender},
+};
+use otap_df_otlp::grpc::OTLPData;
+use otap_df_otlp::proto::opentelemetry::collector::logs::v1::ExportLogsServiceRequest;
+use std::hint::black_box;
+use std::time::{Duration, Instant};
+use tokio::task::LocalSet;
+
+#[global_allocator]
+static GLOBAL: MiMalloc = MiMalloc;
+
+const LARGE_MSG_COUNT: usize = 1_000;
+const PENDING_MESSAGE_SCALES: &[usize] = &[100, 500, 1_000];
+const CLEANUP_SCALES: &[usize] = &[50, 100, 200];
+
+/// Simple mock effect handler that doesn't actually send messages
+struct MockEffectHandler<T> {
+    message_count: std::rc::Rc<std::cell::RefCell<usize>>,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> MockEffectHandler<T> {
+    fn new() -> Self {
+        Self {
+            message_count: std::rc::Rc::new(std::cell::RefCell::new(0)),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn get_count(&self) -> usize {
+        *self.message_count.borrow()
+    }
+
+    async fn send_message(&mut self, _data: T) -> Result<(), otap_df_engine::error::Error<T>> {
+        *self.message_count.borrow_mut() += 1;
+        Ok(())
+    }
+}
+
+/// Creates test OTLP logs data for benchmarking
+fn create_otlp_logs_data() -> OTLPData {
+    let logs_request = ExportLogsServiceRequest::default();
+    OTLPData::Logs(logs_request)
+}
+
+/// Benchmark 1: Message throughput comparison between success path and retry scenarios
+fn bench_message_throughput(c: &mut Criterion) {
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .expect("failed to build Tokio runtime");
+
+    // Pin to a core for consistent measurements
+    let cores = core_affinity::get_core_ids().expect("couldn't get core IDs");
+    let core = cores.iter().last().expect("no cores found");
+    _ = core_affinity::set_for_current(*core);
+
+    let mut group = c.benchmark_group("retry_processor_throughput");
+
+    // Benchmark pure success path (0% retries)
+    let _ = group.throughput(Throughput::Elements(LARGE_MSG_COUNT as u64));
+    let _ = group.bench_function(BenchmarkId::new("success_path", LARGE_MSG_COUNT), |b| {
+        b.to_async(&rt).iter(|| async {
+            let local = LocalSet::new();
```

Comment:
nit: we might be able to create the LocalSet outside the benchmark loop to avoid having the time it takes to call `LocalSet::new()` included in the benchmark time:

```suggrestion
        let local = LocalSet::new();
        b.to_async(&rt).iter(|| async {
```